pkg/reconciler/clusterimagepolicy/clusterimagepolicy.go

```diff
@@ -116,18 +116,60 @@ func (r *Reconciler) FinalizeKind(ctx context.Context, cip *v1alpha1.ClusterImag
 	return nil
 }
 
-// Checks to see if we can deal with format yet. This is missing support
-// for things like Secret resolution, so we can't do those yet. As more things
-// are supported, remove them from here.
-func willItBlend(cip *v1alpha1.ClusterImagePolicy) bool {
-	for _, authority := range cip.Spec.Authorities {
+// inlineSecrets will go through the CIP and try to read the referenced
+// secrets and convert them into inlined data. Makes a copy of the CIP
+// before modifying it and returns the copy.
+func (r *Reconciler) inlineSecrets(ctx context.Context, cip *v1alpha1.ClusterImagePolicy) (*v1alpha1.ClusterImagePolicy, error) {
+	ret := cip.DeepCopy()
+	for _, authority := range ret.Spec.Authorities {
 		if authority.Key != nil && authority.Key.SecretRef != nil {
-			return false
+			if err := r.inlineAndTrackSecret(ctx, ret, authority.Key); err != nil {
+				logging.FromContext(ctx).Errorf("Failed to read secret %q: %w", authority.Key.SecretRef.Name)
+				return nil, err
+			}
 		}
 		if authority.Keyless != nil && authority.Keyless.CAKey != nil &&
 			authority.Keyless.CAKey.SecretRef != nil {
-			return false
+			if err := r.inlineAndTrackSecret(ctx, ret, authority.Keyless.CAKey); err != nil {
+				logging.FromContext(ctx).Errorf("Failed to read secret %q: %w", authority.Keyless.CAKey.SecretRef.Name)
+				return nil, err
+			}
 		}
 	}
-	return true
+	return ret, nil
+}
+
+// inlineSecret will take in a KeyRef and tries to read the Secret, finding the
+// first key from it and will inline it in place of Data and then clear out
+// the SecretRef and return it.
+// Additionally, we set up a tracker so we will be notified if the secret
+// is modified.
+// There's still some discussion about how to handle multiple keys in a secret
+// for now, just grab one from it. For reference, the discussion is here:
+// TODO(vaikas): https://github.com/sigstore/cosign/issues/1573
+func (r *Reconciler) inlineAndTrackSecret(ctx context.Context, cip *v1alpha1.ClusterImagePolicy, keyref *v1alpha1.KeyRef) error {
+	if err := r.tracker.TrackReference(tracker.Reference{
+		APIVersion: "v1",
+		Kind:       "Secret",
+		Namespace:  system.Namespace(),
+		Name:       keyref.SecretRef.Name,
+	}, cip); err != nil {
+		return fmt.Errorf("failed to track changes to secret %q : %w", keyref.SecretRef.Name, err)
+	}
+	secret, err := r.secretlister.Secrets(system.Namespace()).Get(keyref.SecretRef.Name)
+	if err != nil {
+		return err
+	}
+	if len(secret.Data) == 0 {
+		return fmt.Errorf("secret %q contains no data", keyref.SecretRef.Name)
+	}
+	if len(secret.Data) > 1 {
```

Comment:
what about validating that the data field contains a `publicKey` field like here https://github.com/sigstore/cosign/pull/1595/files#diff-f95503af2405ad769f56a5e3ec662f2d57380441e636ba81ffe33b93e00ea4ebR437 ?

Reply:
I'll do that in a followup?
https://github.com/sigstore/cosign/issues/1596